### PR TITLE
[KOGITO-4076] Added dmn-drools-springboot-metrics example

### PR DESCRIPTION
**Pull request by danielezonca**
https://issues.redhat.com/browse/KOGITO-4076

PRs:
- https://github.com/kiegroup/kogito-runtimes/pull/955
- https://github.com/kiegroup/kogito-examples/pull/493

### CODE REVIEW

File: dmn-drools-springboot-metrics/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+ENV JAVA_OPTIONS="-Dquarkus.http.host=0.0.0.0 -Djava.util.logging.manager=org.jboss.logmanager.LogManager"
+
+COPY target/lib/* /deployments/lib/
+COPY target/*-runner.jar /deployments/app.jar
+
+EXPOSE 8080
+EXPOSE 3000
+USER 1001
+
+ENTRYPOINT [ "/deployments/run-java.sh" ]
```
Comment (r00ta) [inline dmn-drools-springboot-metrics/Dockerfile]:
```suggestion
FROM quay.io/kiegroup/kogito-springboot-ubi8:latest

COPY target/*.jar $KOGITO_HOME/bin
```

File: dmn-drools-springboot-metrics/README.md [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+```XML
+<dependency>
+ <groupId>org.kie.kogito</groupId>
+ <artifactId>monitoring-prometheus-quarkus-addon</artifactId>
+</dependency>
+```
+
+### Architecture
+
+Once you compile your Quarkus project, a dashboard for each available endpoint will be generated under the path `target/resources/dashboards/`. You can then inject those grafana dashboards during the deployment of the grafana instance.
```
Comment (r00ta) [inline dmn-drools-springboot-metrics/README.md]:
```suggestion
Once you compile your Springboot project, a dashboard for each available endpoint will be generated under the path `target/resources/dashboards/`. You can then inject those grafana dashboards during the deployment of the grafana instance.
```

**Review by r00ta — APPROVED** [review]
Looks good to me! Thank you very much :+1:

File: dmn-drools-quarkus-metrics/src/test/java/org/kie/kogito/examples/LoanEligibilityTest.java [hunk 3]
```diff
@@ -61,6 +61,7 @@ public void testMetricsLoanEligibility() {
     }
 
     @Test
+    @SuppressWarnings("unchecked")
```
Comment (jiripetrlik) [inline dmn-drools-quarkus-metrics/src/test/java/org/kie/kogito/examples/LoanEligibilityTest.java:64]: Is this related to the topic of this PR - "Added dmn-drools-springboot-metrics example"?

File: dmn-drools-springboot-metrics/README.md [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+### Prerequisites
+ 
+You will need:
+  - Java 11+ installed 
+  - Environment variable JAVA_HOME set accordingly
+  - Maven 3.6.2+ installed
+  
+### How to enable the feature
+
+Import the following dependency in your `pom.xml` if you are using quarkus:
```
Comment (jiripetrlik) [inline dmn-drools-springboot-metrics/README.md]: It seems that there is springboot addon in the snippet bellow.

File: dmn-drools-springboot-metrics/README.md [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+
+```XML
+<dependency>
+ <groupId>org.kie.kogito</groupId>
+ <artifactId>monitoring-prometheus-springboot-addon</artifactId>
+</dependency>
+```
+
+Otherwise, if you are using quarkus import
```
Comment (jiripetrlik) [inline dmn-drools-springboot-metrics/README.md]: Does it make sense to write about Quarkus addon in the Springboot example?

File: dmn-drools-springboot-metrics/src/test/java/org/kie/kogito/examples/GrafanaDockerComposeIT.java [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+import org.junit.jupiter.api.Test;
+import org.testcontainers.containers.DockerComposeContainer;
+import org.testcontainers.containers.wait.strategy.Wait;
+import org.testcontainers.junit.jupiter.Container;
+import org.testcontainers.junit.jupiter.Testcontainers;
+
+import static io.restassured.RestAssured.given;
+import static org.hamcrest.Matchers.hasItem;
+
+@Disabled("Flaky test. Must disable temporarily")
```
Comment (jiripetrlik) [inline dmn-drools-springboot-metrics/src/test/java/org/kie/kogito/examples/GrafanaDockerComposeIT.java]: Can you please provide more info about why is this test failing?

File: process-infinispan-persistence-quarkus/src/test/java/org/acme/deals/DealsRestIT.java [hunk 7]
```diff
@@ -97,6 +97,7 @@ public void testDealsRest() {
     }
 
     @Test
+    @SuppressWarnings("unchecked")
```
Comment (jiripetrlik) [inline process-infinispan-persistence-quarkus/src/test/java/org/acme/deals/DealsRestIT.java:100]: Is this related to this PR?

**Review by jiripetrlik — CHANGES_REQUESTED** [review]
I have few comments to this PR.

File: dmn-drools-springboot-metrics/src/test/java/org/kie/kogito/examples/GrafanaDockerComposeIT.java (hunk 6 above)
Comment (danielezonca) [reply dmn-drools-springboot-metrics/src/test/java/org/kie/kogito/examples/GrafanaDockerComposeIT.java]: This example is a carbon copy of the existing `dmn-drools-quarkus-metrics` that is currently disabled on master ( see [link](https://github.com/kiegroup/kogito-examples/blob/master/dmn-drools-quarkus-metrics/src/test/java/org/kie/kogito/examples/GrafanaDockerComposeIT.java#L31) ).

@r00ta is fixing it and re-enabling with this PR
https://github.com/kiegroup/kogito-examples/pull/495
and then I can apply the same changes.

:thinking: I think it is better to wait @r00ta PR to be merged before this one so I can enable the test. Wdyt?

File: dmn-drools-quarkus-metrics/src/test/java/org/kie/kogito/examples/LoanEligibilityTest.java (hunk 3 above)
Comment (danielezonca) [reply dmn-drools-quarkus-metrics/src/test/java/org/kie/kogito/examples/LoanEligibilityTest.java:64]: Nope, it is just to remove the warning from the log, I don't know if it makes sense to have a independent PR just for this

File: process-infinispan-persistence-quarkus/src/test/java/org/acme/deals/DealsRestIT.java (hunk 7 above)
Comment (danielezonca) [reply process-infinispan-persistence-quarkus/src/test/java/org/acme/deals/DealsRestIT.java:100]: Nope, it is just to remove the warning from the log, I don't know if it makes sense to have a independent PR just for this

File: dmn-drools-springboot-metrics/README.md (hunk 5 above)
Comment (danielezonca) [reply dmn-drools-springboot-metrics/README.md]: Done

File: dmn-drools-springboot-metrics/README.md (hunk 4 above)
Comment (danielezonca) [reply dmn-drools-springboot-metrics/README.md]: Done

File: dmn-drools-springboot-metrics/docker-compose.yml [hunk 8]
```diff
@@ -0,0 +1,28 @@
+version: '2'
+
+services:
+  hello:
+    build: ./
```
Comment (r00ta) [inline dmn-drools-springboot-metrics/docker-compose.yml]:
```suggestion
    image: org.kie.kogito.examples/dmn-drools-sprintboot-metrics:1.0
```

File: dmn-drools-springboot-metrics/pom.xml [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+                    <include>docker-compose.yml</include>
+                    <include>docker-compose/</include>
+                  </includes>
+                </resource>
+              </resources>
+            </configuration>
+          </execution>
+        </executions>
+      </plugin>
+    </plugins>
```
Comment (r00ta) [inline dmn-drools-springboot-metrics/pom.xml:187]:
```suggestion
      <plugin>
        <groupId>org.codehaus.mojo</groupId>
        <artifactId>exec-maven-plugin</artifactId>
        <executions>
          <execution>
            <id>docker-build</id>
            <phase>pre-integration-test</phase>
            <goals>
              <goal>exec</goal>
            </goals>
            <configuration>
              <executable>docker</executable>
              <workingDirectory>${project.basedir}</workingDirectory>
              <arguments>
                <argument>build</argument>
                <argument>-t</argument>
                <argument>${project.groupId}/${project.artifactId}:1.0</argument>
                <argument>.</argument>
              </arguments>
            </configuration>
          </execution>
        </executions>
      </plugin>
    </plugins>
```

File: dmn-drools-springboot-metrics/README.md [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+
+This example demonstrates how to enable and consume the runtime metrics monitoring feature in Kogito. 
+
+### Prerequisites
+ 
+You will need:
+  - Java 11+ installed 
+  - Environment variable JAVA_HOME set accordingly
+  - Maven 3.6.2+ installed
+
```
Comment (r00ta) [inline dmn-drools-springboot-metrics/README.md]:
```suggestion
    - Docker 19+ (only if you want to run the integration tests and/or you want to use the `docker-compose` script provided in this example).

```

File: dmn-drools-springboot-metrics/README.md [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+   * if the output of the decision is a number, the graph contains the quantiles for that metric (on a sliding window of 3 minutes).
+   * If the output is a boolean or a string, the graph contains the number of occurrences for each output (10 minutes average).
+![Screenshot from 2020-05-19 15-19-48](https://user-images.githubusercontent.com/18282531/82339828-c71c2c80-99ee-11ea-85b6-b5d4a0337f0b.png)
+
+You can use these default dashboards, or you can personalize them and use your custom dashboards.
+
+### Compile and Run in Local Dev Mode
+
+It is possible to use `docker-compose` to demonstrate how to inject the generated dashboards in the volume of the grafana container:
+1. Run `mvn clean package` to build the project and generate dashboards.
```
Comment (r00ta) [inline dmn-drools-springboot-metrics/README.md]:
```suggestion
1. Run `mvn clean package` to build the project and generate dashboards.  A docker image tagged `org.kie.kogito.examples/dmn-drools-sprintboot-metrics-example:1.0` will be built (docker must be installed on your system).
```

**Review by r00ta — COMMENTED** [review]
Applying the same changes I did here (https://github.com/kiegroup/kogito-examples/pull/520) so to build the docker image just before the integration tests are run

File: dmn-drools-springboot-metrics/docker-compose.yml [hunk 12]
```diff
@@ -0,0 +1,27 @@
+version: '2'
+
+services:
+  hello:
+    image: org.kie.kogito.examples/dmn-drools-sprintboot-metrics:1.0
```
Comment (r00ta) [inline dmn-drools-springboot-metrics/docker-compose.yml]:
```suggestion
    image: org.kie.kogito.examples/dmn-drools-springboot-metrics:1.0
```

File: dmn-drools-springboot-metrics/README.md [hunk 13]
```diff
@@ -0,0 +1,90 @@
...
+   * if the output of the decision is a number, the graph contains the quantiles for that metric (on a sliding window of 3 minutes).
+   * If the output is a boolean or a string, the graph contains the number of occurrences for each output (10 minutes average).
+![Screenshot from 2020-05-19 15-19-48](https://user-images.githubusercontent.com/18282531/82339828-c71c2c80-99ee-11ea-85b6-b5d4a0337f0b.png)
+
+You can use these default dashboards, or you can personalize them and use your custom dashboards.
+
+### Compile and Run in Local Dev Mode
+
+It is possible to use `docker-compose` to demonstrate how to inject the generated dashboards in the volume of the grafana container:
+1. Run `mvn clean package` to build the project and generate dashboards.  A docker image tagged `org.kie.kogito.examples/dmn-drools-sprintboot-metrics-example:1.0` will be built (docker must be installed on your system).
```
Comment (r00ta) [inline dmn-drools-springboot-metrics/README.md]:
```suggestion
1. Run `mvn clean package` to build the project and generate dashboards.  A docker image tagged `org.kie.kogito.examples/dmn-drools-springboot-metrics-example:1.0` will be built (docker must be installed on your system).
```

**Review by r00ta — APPROVED** [review]

**Comment by danielezonca** [comment]
jenkins retest this please

**Comment by r00ta** [comment]
Jenkins retest this please

File: dmn-drools-springboot-metrics/src/test/java/org/kie/kogito/examples/GrafanaDockerComposeIT.java (hunk 6 above)
Comment (danielezonca) [reply dmn-drools-springboot-metrics/src/test/java/org/kie/kogito/examples/GrafanaDockerComposeIT.java]: Done, now is enabled :)

**Review by jiripetrlik — APPROVED** [review]
Thank you for the changes. Looks good to me now.
